File: app/src/main/java/com/mab/adminpanel/QRScannerActivity.java
package com.mab.adminpanel;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.Toast;
import com.google.zxing.Result;
import me.dm7.barcodescanner.zxing.ZXingScannerView;

import static android.support.v4.content.PermissionChecker.PERMISSION_GRANTED;

public class QRScannerActivity extends AppCompatActivity implements ZXingScannerView.ResultHandler {
    private String TAG = QRScannerActivity.class.getSimpleName();
    private int permsRequestCode = 200;
    private String[] perms = new String[]{Manifest.permission.CAMERA};

    private ZXingScannerView qr_view_id;

    @Override
    public void handleResult(Result result) {
        Log.d(TAG, "handleResult" + result.getText());
        Intent intent = new Intent();
        intent.putExtra(MConstants.QR_RESULT_INTENT_FILTER, result.getText());
        setResult(MConstants.QR_RESULT_INTENT_RESULT, intent);
        finish();
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_qrscanner);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(perms, permsRequestCode);
        }
        qr_view_id = findViewById(R.id.qr_view_id);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.CAMERA) != PERMISSION_GRANTED) {
                Toast.makeText(this, "Need permission", Toast.LENGTH_SHORT).show();
                return;
            }
        }
        gotPermissionOfCameraLocation();
    }

    @Override
    protected void onPause() {
        super.onPause();
        qr_view_id.stopCamera();
    }

    private void gotPermissionOfCameraLocation() {
        qr_view_id.setResultHandler(this);
        qr_view_id.startCamera();
    }


    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        Log.d(TAG, "onRequestPermissionsResult");
        switch (requestCode) {
            case 200: {
                boolean cameraAccepted = grantResults[0] == PackageManager.PERMISSION_GRANTED;
                if (cameraAccepted) {
                    gotPermissionOfCameraLocation();
                } else {
                    Toast.makeText(this, "Need permission", Toast.LENGTH_SHORT).show();
                }
            }
        }
    }
}

File: app/src/main/java/com/mab/adminpanel/QRScannerActivity.kt
package com.mab.adminpanel

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.os.Build
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View.GONE
import android.view.View.VISIBLE
import com.google.zxing.Result
import kotlinx.android.synthetic.main.activity_qrscanner.*
import me.dm7.barcodescanner.zxing.ZXingScannerView


class QRScannerActivity : AppCompatActivity(), ZXingScannerView.ResultHandler {
    private val TAG = QRScannerActivity::class.java.simpleName
    private val permsRequestCode = 200
    private val perms = arrayOf(Manifest.permission.CAMERA)


    override fun handleResult(result: Result?) {
        Log.d(TAG, "handleResult ${result?.text}")
        val intent = Intent()
        intent.putExtra(MConstants.QR_RESULT_INTENT_FILTER, result?.text)
        setResult(MConstants.QR_RESULT_INTENT_RESULT, intent)
        finish()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_qrscanner)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            enable_button.setOnClickListener {
                requestPermissions(perms, permsRequestCode)
            }
        }
        initListeners()
    }

    private fun initListeners() {
        cancel_button.setOnClickListener {
            finish()
        }

    }

    public override fun onResume() {
        super.onResume()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.CAMERA) != PERMISSION_GRANTED) {
                notHavePermissionForCameraLocation()
                return
            }
        }
        gotPermissionOfCameraLocation()

    }

    private fun notHavePermissionForCameraLocation() {
        info_view_id.visibility = VISIBLE
        qr_layout.visibility = GONE
    }

    private fun gotPermissionOfCameraLocation() {
        info_view_id.visibility = GONE
        qr_layout.visibility = VISIBLE
        qr_view_id.setResultHandler(this)
        qr_view_id.startCamera()
    }

    public override fun onPause() {
        super.onPause()
        qr_view_id.stopCamera()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        Log.d(TAG, "onRequestPermissionsResult")
        when (requestCode) {
            200 -> {
                val cameraAccepted = grantResults[0] == PackageManager.PERMISSION_GRANTED
                if (cameraAccepted) {
                    gotPermissionOfCameraLocation()
                } else {
                    notHavePermissionForCameraLocation()
                }
            }
        }
    }
}

